alsa: allow 'sound.mediaKeys.enable' to work even when 'hardware.pulseaudio.enable' is used

nrdxp opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
This resolves #66854 by adding new functionality to actkbd, allowing it to selectively deploy a systemd user service when a special 'userMode' toggle is set in the bindings attribute set. This does not work globally, but per binding so you can still have system-wide bindings for those that work fine that way, and user session bindings for any program which requires user permissions. I have split the config files for each instance so that the same binding will not be called by both services.

Additionally, I have modified 'sound.mediaKeys.enable' to take advantage of this new feature so that bindings for volume control will still work even when 'hardware.pulseaudio.enable' is set.

I imagine many users of NixOS are advanced users that use a lightweight window manager and not a full DE, so this should benefit anybody in that situation to have volume keys that 'just work' even when using a Wayland session or no graphical session at all.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

nrdxp commented:
I decided to change the `volumeStep` option to simplify the implentation. If you look at the 3rd commit you can see the original `volumeStep` model working using `pactl`, however to use `pactl` and target the default sink or source I have to do some needless command line work while `pamixer` targets the default sink or source automatically but it doesn't allow decibel calculations. 

In commit 4, I opted for the simpler implementation of `pamixer` and changed `volumeStep` to an int since moving volume by percentage points seems to be a decent compromise. Users can still change their `volumeStep` they just can't calculate by decibel or direct hw values anymore. As an added bonus, users can decided whether or not they'd like to allow pulse to go beyond 100% volume with the new `allowBoost` toggle.

Technically we can use either one of these implementations; they both work fine. I'd just rather call a single process rather than a shell calling a process that calls a process.

Commit:
actkbd: add ability to deploy as user service

Some processes require user level permissions to interact with them,
E.g. pulseaudio. Allow users to make keybindings for them.

* add new 'userMode' toggle to bindings; default to 'false'
* add 'ENV{SYSTEMD_USER_WANTS}' to udev rule
* only launch user service instances if there are relevant bindings
* add 'start-actkbd' user service to initialize actkbd instances

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/55

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review-may-2019/3032/47

Commit:
alsa: import 'mediaKeys' to work with pulseaudio

Resolves https://github.com/NixOS/nixpkgs/issues/66854

Media keys are audio server agnostic and 'just work'

* declare new keybinds for pulseaudio using pamixer
* set new keybinds if pulseaudio is enabled otherwise use original binds

nrdxp commented:
Just did a force push to use pactl instead of pamixer. I didn't realize that you could use @DEFAULT_SINK@ for pactl until today (its not in the manpage). 

This simplifies the PR a bit as I no longer needed to modify the volumeStep type so this should be much less intrusive. Should work with all existing configurations. I everything again just to be sure it still works and its good to go.

Any takers?

rileyinman commented:
Any reason this was closed besides inactivity? This seems like a good option to have available for pulseaudio users.

ptman commented on nixos/modules/services/hardware/actkbd.nix:
```diff
@@ -6,18 +6,47 @@ let
+      ConditionPathExists = "%I";
+    };
+    serviceConfig = {
+      Type = "forking";
```
I believe forking could be changed to simple and -D removed from the ExecStart -line.
